File: PrimeGenerator.java
// PrimeGenerator.java - Генератор простых чисел на Java (CLI)
import java.io.*;
import java.util.*;

public class PrimeGenerator {
    private static Scanner scanner = new Scanner(System.in);

    public static List<Integer> sieveOfEratosthenes(int n) {
        if (n < 2) return new ArrayList<>();
        boolean[] isPrime = new boolean[n + 1];
        Arrays.fill(isPrime, true);
        isPrime[0] = isPrime[1] = false;
        for (int i = 2; i * i <= n; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i <= n; i++) {
            if (isPrime[i]) primes.add(i);
        }
        return primes;
    }

    public static boolean isPrimeTrial(int n) {
        if (n < 2) return false;
        if (n == 2) return true;
        if (n % 2 == 0) return false;
        for (int i = 3; i * i <= n; i += 2) {
            if (n % i == 0) return false;
        }
        return true;
    }

    public static List<Integer> primesInRange(int a, int b, String method) {
        if (a < 2) a = 2;
        if (a > b) return new ArrayList<>();
        if (method.equals("sieve")) {
            List<Integer> all = sieveOfEratosthenes(b);
            List<Integer> res = new ArrayList<>();
            for (int p : all) {
                if (p >= a) res.add(p);
            }
            return res;
        } else {
            List<Integer> res = new ArrayList<>();
            for (int p = a; p <= b; p++) {
                if (isPrimeTrial(p)) res.add(p);
            }
            return res;
        }
    }

    public static Map<String, Object> stats(List<Integer> primes) {
        Map<String, Object> st = new HashMap<>();
        int count = primes.size();
        st.put("count", count);
        if (count == 0) {
            st.put("sum", 0);
            st.put("avg", 0.0);
            st.put("min", 0);
            st.put("max", 0);
            return st;
        }
        long sum = 0;
        for (int p : primes) sum += p;
        st.put("sum", sum);
        st.put("avg", (double) sum / count);
        st.put("min", primes.get(0));
        st.put("max", primes.get(count - 1));
        return st;
    }

    public static void exportToFile(String filename, List<Integer> primes) throws IOException {
        try (PrintWriter pw = new PrintWriter(new FileWriter(filename))) {
            for (int p : primes) {
                pw.println(p);
            }
        }
    }

    public static void interactive() {
        System.out.println("🔢 Генератор простых чисел (интерактивный)");
        while (true) {
            System.out.println("\nВыберите действие:");
            System.out.println("1. Сгенерировать простые числа до N");
            System.out.println("2. Сгенерировать простые числа в диапазоне [A, B]");
            System.out.println("3. Проверить число на простоту");
            System.out.println("4. Выход");
            System.out.print("Ваш выбор: ");
            String choice = scanner.nextLine().trim();
            if (choice.equals("4")) break;
            else if (choice.equals("1")) {
                System.out.print("Введите N: ");
                int n = Integer.parseInt(scanner.nextLine().trim());
                long start = System.currentTimeMillis();
                List<Integer> primes = sieveOfEratosthenes(n);
                long elapsed = System.currentTimeMillis() - start;
                Map<String, Object> st = stats(primes);
                System.out.printf("Простых чисел: %d, Сумма: %d, Среднее: %.2f, Время: %.3f сек.\n",
                        st.get("count"), st.get("sum"), st.get("avg"), elapsed / 1000.0);
                System.out.print("Показать список? (y/n): ");
                String show = scanner.nextLine().trim().toLowerCase();
                if (show.equals("y")) {
                    System.out.println(primes);
                }
                System.out.print("Экспортировать в файл? (y/n): ");
                String exp = scanner.nextLine().trim().toLowerCase();
                if (exp.equals("y")) {
                    System.out.print("Имя файла: ");
                    String fname = scanner.nextLine().trim();
                    if (fname.isEmpty()) fname = "primes.txt";
                    try {
                        exportToFile(fname, primes);
                        System.out.println("Сохранено в " + fname);
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
            } else if (choice.equals("2")) {
                System.out.print("Введите A: ");
                int a = Integer.parseInt(scanner.nextLine().trim());
                System.out.print("Введите B: ");
                int b = Integer.parseInt(scanner.nextLine().trim());
                System.out.print("Метод (sieve/trial): ");
                String method = scanner.nextLine().trim().toLowerCase();
                if (!method.equals("sieve") && !method.equals("trial")) method = "sieve";
                long start = System.currentTimeMillis();
                List<Integer> primes = primesInRange(a, b, method);
                long elapsed = System.currentTimeMillis() - start;
                Map<String, Object> st = stats(primes);
                System.out.printf("Простых чисел: %d, Сумма: %d, Среднее: %.2f, Время: %.3f сек.\n",
                        st.get("count"), st.get("sum"), st.get("avg"), elapsed / 1000.0);
                System.out.print("Показать список? (y/n): ");
                String show = scanner.nextLine().trim().toLowerCase();
                if (show.equals("y")) {
                    System.out.println(primes);
                }
                System.out.print("Экспортировать в файл? (y/n): ");
                String exp = scanner.nextLine().trim().toLowerCase();
                if (exp.equals("y")) {
                    System.out.print("Имя файла: ");
                    String fname = scanner.nextLine().trim();
                    if (fname.isEmpty()) fname = "primes.txt";
                    try {
                        exportToFile(fname, primes);
                        System.out.println("Сохранено в " + fname);
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
            } else if (choice.equals("3")) {
                System.out.print("Введите число: ");
                int num = Integer.parseInt(scanner.nextLine().trim());
                boolean isPrime = isPrimeTrial(num);
                System.out.printf("%d %s простым.\n", num, isPrime ? "является" : "не является");
            } else {
                System.out.println("Неверный выбор.");
            }
        }
    }

    public static void main(String[] args) {
        if (args.length > 0) {
            // CLI
            int max = 0, min = 2;
            String method = "sieve";
            String exportFile = null;
            boolean statsFlag = false;
            for (int i = 0; i < args.length; i++) {
                switch (args[i]) {
                    case "--max": max = Integer.parseInt(args[++i]); break;
                    case "--min": min = Integer.parseInt(args[++i]); break;
                    case "--method": method = args[++i]; break;
                    case "--export": exportFile = args[++i]; break;
                    case "--stats": statsFlag = true; break;
                }
            }
            if (max > 0) {
                long start = System.currentTimeMillis();
                List<Integer> primes;
                if (method.equals("sieve")) {
                    primes = sieveOfEratosthenes(max);
                } else {
                    primes = primesInRange(min, max, method);
                }
                long elapsed = System.currentTimeMillis() - start;
                if (statsFlag) {
                    Map<String, Object> st = stats(primes);
                    System.out.printf("Количество: %d, Сумма: %d, Среднее: %.2f, Мин: %d, Макс: %d\n",
                            st.get("count"), st.get("sum"), st.get("avg"), st.get("min"), st.get("max"));
                    System.out.printf("Время: %.3f сек.\n", elapsed / 1000.0);
                } else {
                    System.out.println(primes);
                }
                if (exportFile != null) {
                    try {
                        exportToFile(exportFile, primes);
                        System.out.println("Экспортировано в " + exportFile);
                    } catch (IOException e) {
                        System.out.println("Ошибка экспорта: " + e.getMessage());
                    }
                }
            } else {
                System.out.println("Использование: java PrimeGenerator --max N [--min M] [--method sieve|trial] [--export file] [--stats]");
            }
        } else {
            interactive();
        }
    }
}

File: prime_generator.rs
// prime_generator.rs - Генератор простых чисел на Rust (CLI)
use std::io::{self, Write};
use std::time::Instant;
use std::env;
use std::fs::File;
use std::io::prelude::*;

fn sieve_of_eratosthenes(n: usize) -> Vec<usize> {
    if n < 2 {
        return vec![];
    }
    let mut is_prime = vec![true; n + 1];
    is_prime[0] = false;
    is_prime[1] = false;
    let limit = (n as f64).sqrt() as usize;
    for i in 2..=limit {
        if is_prime[i] {
            let mut j = i * i;
            while j <= n {
                is_prime[j] = false;
                j += i;
            }
        }
    }
    (2..=n).filter(|&i| is_prime[i]).collect()
}

fn is_prime_trial(n: usize) -> bool {
    if n < 2 { return false; }
    if n == 2 { return true; }
    if n % 2 == 0 { return false; }
    let limit = (n as f64).sqrt() as usize;
    for i in (3..=limit).step_by(2) {
        if n % i == 0 {
            return false;
        }
    }
    true
}

fn primes_in_range(a: usize, b: usize, method: &str) -> Vec<usize> {
    let start = if a < 2 { 2 } else { a };
    if start > b {
        return vec![];
    }
    if method == "sieve" {
        let all = sieve_of_eratosthenes(b);
        all.into_iter().filter(|&p| p >= start).collect()
    } else {
        (start..=b).filter(|&p| is_prime_trial(p)).collect()
    }
}

fn stats(primes: &[usize]) -> (usize, usize, f64, usize, usize) {
    let count = primes.len();
    if count == 0 {
        return (0, 0, 0.0, 0, 0);
    }
    let sum: usize = primes.iter().sum();
    let avg = sum as f64 / count as f64;
    let min = primes[0];
    let max = primes[count - 1];
    (count, sum, avg, min, max)
}

fn export_to_file(filename: &str, primes: &[usize]) -> Result<(), std::io::Error> {
    let mut file = File::create(filename)?;
    for p in primes {
        writeln!(file, "{}", p)?;
    }
    Ok(())
}

fn read_line(prompt: &str) -> String {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Ошибка ввода");
    input.trim().to_string()
}

fn interactive() {
    println!("🔢 Генератор простых чисел (интерактивный)");
    loop {
        println!("\nВыберите действие:");
        println!("1. Сгенерировать простые числа до N");
        println!("2. Сгенерировать простые числа в диапазоне [A, B]");
        println!("3. Проверить число на простоту");
        println!("4. Выход");
        let choice = read_line("Ваш выбор: ");
        match choice.as_str() {
            "4" => break,
            "1" => {
                let n: usize = read_line("Введите N: ").parse().expect("Ошибка ввода");
                let start = Instant::now();
                let primes = sieve_of_eratosthenes(n);
                let elapsed = start.elapsed().as_secs_f64();
                let (count, sum, avg, min, max) = stats(&primes);
                println!("Простых чисел: {}, Сумма: {}, Среднее: {:.2}, Время: {:.4} сек.", count, sum, avg, elapsed);
                let show = read_line("Показать список? (y/n): ");
                if show.to_lowercase() == "y" {
                    println!("{:?}", primes);
                }
                let exp = read_line("Экспортировать в файл? (y/n): ");
                if exp.to_lowercase() == "y" {
                    let fname = read_line("Имя файла: ");
                    let fname = if fname.is_empty() { "primes.txt".to_string() } else { fname };
                    if let Err(e) = export_to_file(&fname, &primes) {
                        println!("Ошибка сохранения: {}", e);
                    } else {
                        println!("Сохранено в {}", fname);
                    }
                }
            }
            "2" => {
                let a: usize = read_line("Введите A: ").parse().expect("Ошибка ввода");
                let b: usize = read_line("Введите B: ").parse().expect("Ошибка ввода");
                let method = read_line("Метод (sieve/trial): ");
                let method = if method == "trial" { "trial" } else { "sieve" };
                let start = Instant::now();
                let primes = primes_in_range(a, b, method);
                let elapsed = start.elapsed().as_secs_f64();
                let (count, sum, avg, min, max) = stats(&primes);
                println!("Простых чисел: {}, Сумма: {}, Среднее: {:.2}, Время: {:.4} сек.", count, sum, avg, elapsed);
                let show = read_line("Показать список? (y/n): ");
                if show.to_lowercase() == "y" {
                    println!("{:?}", primes);
                }
                let exp = read_line("Экспортировать в файл? (y/n): ");
                if exp.to_lowercase() == "y" {
                    let fname = read_line("Имя файла: ");
                    let fname = if fname.is_empty() { "primes.txt".to_string() } else { fname };
                    if let Err(e) = export_to_file(&fname, &primes) {
                        println!("Ошибка сохранения: {}", e);
                    } else {
                        println!("Сохранено в {}", fname);
                    }
                }
            }
            "3" => {
                let num: usize = read_line("Введите число: ").parse().expect("Ошибка ввода");
                let is_prime = is_prime_trial(num);
                println!("{} {} простым.", num, if is_prime { "является" } else { "не является" });
            }
            _ => println!("Неверный выбор."),
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() > 1 {
        let mut max = 0;
        let mut min = 2;
        let mut method = "sieve".to_string();
        let mut export_file = String::new();
        let mut stats_flag = false;
        let mut i = 1;
        while i < args.len() {
            match args[i].as_str() {
                "--max" => {
                    max = args[i + 1].parse().expect("Ошибка");
                    i += 2;
                }
                "--min" => {
                    min = args[i + 1].parse().expect("Ошибка");
                    i += 2;
                }
                "--method" => {
                    method = args[i + 1].clone();
                    i += 2;
                }
                "--export" => {
                    export_file = args[i + 1].clone();
                    i += 2;
                }
                "--stats" => {
                    stats_flag = true;
                    i += 1;
                }
                _ => {
                    i += 1;
                }
            }
        }
        if max > 0 {
            let start = Instant::now();
            let primes = if method == "sieve" {
                sieve_of_eratosthenes(max)
            } else {
                primes_in_range(min, max, &method)
            };
            let elapsed = start.elapsed().as_secs_f64();
            if stats_flag {
                let (count, sum, avg, min, max) = stats(&primes);
                println!("Количество: {}, Сумма: {}, Среднее: {:.2}, Мин: {}, Макс: {}", count, sum, avg, min, max);
                println!("Время: {:.4} сек.", elapsed);
            } else {
                println!("{:?}", primes);
            }
            if !export_file.is_empty() {
                if let Err(e) = export_to_file(&export_file, &primes) {
                    println!("Ошибка экспорта: {}", e);
                } else {
                    println!("Экспортировано в {}", export_file);
                }
            }
        } else {
            println!("Использование: prime_generator --max N [--min M] [--method sieve|trial] [--export file] [--stats]");
        }
    } else {
        interactive();
    }
}
